### PR TITLE
feat(common): introduce async stack trace

### PR DESCRIPTION
**Pull request by BugenZhao**
I hereby agree to the terms of the [Singularity Data, Inc. Contributor License Agreement](https://gist.github.com/skyzh/0663682a70b0edde7ae991492f2314cb#file-s9y_cla).

## What's changed and what's your intention?

This PR introduces a brand new tracing of _async stack trace_, which allows one to check where all tasks are pending and why it's pending. Note that multiple futures can be polled in a single task with `select` and `join`, so the pending status snapshot is actually a tree instead of a stack.

### Samples

To make it intuitive, here're some samples.

- The chain actor has been running for 1s. The rearranged chain steals a barrier 427ms ago, while the batch query executor on the other arm has waited for the epoch for 829ms and is still unfinished yet. Since there're multiple iterators for different vnodes and we `try_join_all` on them, they're in the same level in the tree.
```
Actor 5170 [1.005471416s]
  Epoch 2771234005188608 [427.634708ms]
    RearrangedChain (actor 5170, executor 181) [427.632666ms]
      BatchQuery (actor 5170, executor 180) [829.7615ms]
        store_wait_epoch [829.638ms]
        store_wait_epoch [829.655416ms]
        store_wait_epoch [829.576041ms]
        ...
      ReceiverExecutor (actor 5170, executor 0) [427.644333ms]
        LocalInput (actor 4111) [427.642458ms]
```

- The source executor received a barrier of `2777905903042560` 156ms ago, while there are no new chunks received for 8s.
```
Actor 8 [8.000397291s]
  Epoch 2777905903042560 [156.222541ms]
    SourceExecutor 800000001 (actor 8, executor 1) [156.219166ms]
      source_recv_barrier [156.218875ms]
      source_recv_chunk [!!! 7.996535125s]
```

- This actor tried to dispatch the barrier to the downstream, but has been back-pressured for 7.5s! Will there be some problems in actor 68?
```
Actor 122 [13.001253875s]
  Epoch 2776266980392960 [!!! 7.793291666s]
    dispatch_barrier [!!! 7.550658708s]
      LocalOutput (actor 68) [!!! 7.5506555s]
```

- For actor 68, there's only one arm in the trace of hash join's barrier aligning, which means the other side has been ready for a while. However, we've waited for 11s on the remote inputs and no messages have been received. There must be some issues.
The epoch number in this actor `2667` is also smaller than the actor 122 `2669`, showing that the concurrent checkpoint is stuck. (A real case from [here](https://github.com/singularity-data/risingwave/issues/4270#issuecomment-1202273243) 😄)
```
Actor 68 [14.501581875s]
  Epoch 2776266751148032 [!!! 11.714676875s]
    ProjectExecutor 4400000082 (actor 68, executor 130) [!!! 11.714678s]
      FilterExecutor 4400000080 (actor 68, executor 128) [!!! 11.714679208s]
        HashJoinExecutor 440000007E (actor 68, executor 126) [!!! 11.714680208s]
          hash_join_barrier_align [!!! 11.714681125s]
            MergeExecutor (actor 68, executor 124) [!!! 11.714596916s]
              RemoteInput (actor 141) [!!! 11.714595583s]
              RemoteInput (actor 142) [!!! 11.714596458s]
              RemoteInput (actor 143) [!!! 11.714597083s]
              ...
```

### Usages

To get the traces printed, one may **launch the cluster with env-var `RW_ASYNC_STACK_TRACE` set** and use `risedev ctl stream trace [--actor-id 233]` to find a recent captured report. You may also print it periodically in the logs by commenting out this line.

https://github.com/singularity-data/risingwave/blob/bc44aa526fced1ddfab523f6d133b05c4ff48a7c/src/compute/src/server.rs#L199-L205

The tags must be attached manually (unless we hack the future compiling in `rustc`). Luckily, it's easy enough by adding a chained call of `.stack_trace(..)`. `&'static str` and `String` are both acceptable here, while the former one may perform better.

https://github.com/singularity-data/risingwave/blob/bc44aa526fced1ddfab523f6d133b05c4ff48a7c/src/storage/src/monitor/monitored_store.rs#L232-L236

### Implementations

Writing a double-linked tree in Rust is hard and ugly, not to mention that we must take care of future pending, ready, aborting, and sending to other tasks. I finally decided to implement this with the arena tree, which can be both cleaner and more efficient. For the tracing part, if you're interested, check `src/utils/stack_trace`.

### Limitations

Since we need to manipulate the tree every time we polled a future, this seems significantly hurt the performance with dev profile. That's why we disable it by default. May further investigate it later.

## Checklist

- [x] I have written necessary rustdoc comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
- An enhancement of #3969
- Might be helpful for #4270

### CODE REVIEW

File: src/storage/src/hummock/block_cache.rs [hunk 1]
```diff
@@ -135,6 +136,7 @@ impl BlockCache {
                     Ok((block, len))
                 }
             })
+            .stack_trace("block_cache_lookup")
```
Comment (skyzh) [inline src/storage/src/hummock/block_cache.rs:139]: I think this is not a "stack_trace". It's like a tracing span instead. I can do the renaming later when I introduce minitrace to the system. At that time, we might have something like `.rw_span` which creates both minitrace span and async stack trace info,

**Review by skyzh — APPROVED** [review]
Rest LGTM, good work!

File: src/compute/src/server.rs [hunk 2]
```diff
@@ -196,6 +196,14 @@ pub async fn compute_node_serve(
         state_store,
     );
 
+    // Generally, one may use `risedev ctl stream trace` to manually get the trace reports. However,
+    // if this is not the case, we can use the following command to get it printed into the logs
+    // periodically.
+    //
+    // Comment out the following line to enable.
+    #[cfg(any())]
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/compute/src/server.rs:205]: Agree. I've left a TODO here for now. We may introduce a more complete and unified configuration for all kinds of tracings.

File: src/ctl/src/cmd_impl/stream/trace.rs [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+    let workers = meta_client.get_cluster_info().await?.worker_nodes;
+    let compute_nodes = workers
+        .into_iter()
+        .filter(|w| w.r#type() == WorkerType::ComputeNode);
+
+    let clients = StreamClientPool::new(u64::MAX);
+
+    let mut all_traces = BTreeMap::new();
+
+    // FIXME: the compute node may not be accessible from the our network with their listen
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/ctl/src/cmd_impl/stream/trace.rs]: Here we assume the streaming service of compute nodes is also accessible with `risectl`. However, it's possible that the cluster is running in an internal network and only the meta service is exposed, so we may fail to connect to the compute nodes. Left a more detailed explanation here.🥵

**Review by wangrunji0408 — APPROVED** [review]
LGTM!

Next step we can add some unit tests to make sure the reported traces are equal to our expectations. I'm also looking forward to seeing it on crates.io! 🥰

File: src/utils/async_stack_trace/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+mod manager;
+
+pub use manager::{StackTraceManager, StackTraceReport, TraceReporter};
+
+/// A cheaply-cloneable span string.
+#[derive(Debug, Clone)]
+pub enum SpanValue {
+    Slice(&'static str),
+    Shared(Arc<String>),
+}
```
[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/utils/async_stack_trace/src/lib.rs:41]: I wanna make this struct cheap to clone, so `Arc<String>` is used. For example, we need to trace the Future `actor.next()` multiple times with the same span `Actor 233`, and they'd better share a span instance.

BTW, I've just realized that constructing an `Arc<str>` from `&'static str` or `String` both leads to allocation. 🥵

[the remark this answers is not in the document]
Comment (TennyZhuang) [reply src/utils/async_stack_trace/src/lib.rs:41]: So is it Cow<Arc<str>> ?

[the remark this answers is not in the document]
Comment (wangrunji0408) [reply src/utils/async_stack_trace/src/lib.rs:41]: Sorry... I used to think that cloning `Cow::Owned` only borrows a reference, actually it clones the data! 😰

[the remark this answers is not in the document]
Comment (BugenZhao) [reply src/utils/async_stack_trace/src/lib.rs:41]: > So is it Cow<Arc> ?

`Cow` only accepts a single borrowed type, the owned type must be `<B as ToOwned>::Owned`. 🥵

**Pull request merged**